src/transfer-manager.ts

```diff
@@ -69,6 +84,197 @@ export interface DownloadFileInChunksOptions {
   validation?: 'crc32c' | false;
 }
 
+export interface UploadFileInChunksOptions {
+  concurrencyLimit?: number;
+  chunkSizeBytes?: number;
+  uploadName?: string;
+  maxQueueSize?: number;
+  uploadId?: string;
+  partsMap?: Map<number, string>;
+}
+
+export interface MultiPartUploadHelper {
+  bucket: Bucket;
+  fileName: string;
+  uploadId?: string;
+  partsMap?: Map<number, string>;
+  initiateUpload(): Promise<void>;
+  uploadPart(partNumber: number, chunk: Buffer): Promise<void>;
+  completeUpload(): Promise<void>;
+}
+
+export type MultiPartHelperGenerator = (
+  bucket: Bucket,
+  fileName: string,
+  uploadId?: string,
+  partsMap?: Map<number, string>
+) => MultiPartUploadHelper;
+
+const defaultMultPartGenerator: MultiPartHelperGenerator = (
+  bucket,
+  fileName,
+  uploadId,
+  partsMap
+) => {
+  return new XMLMultiPartUploadHelper(bucket, fileName, uploadId, partsMap);
+};
+
+export class MultiPartUploadError extends Error {
+  private uploadId: string;
+  private partsMap: Map<number, string>;
+
+  constructor(
+    message: string,
+    uploadId: string,
+    partsMap: Map<number, string>
+  ) {
+    super(message);
+    this.uploadId = uploadId;
+    this.partsMap = partsMap;
+  }
+}
+/**
+ * Class representing an implementation of MPU in the XML API. This class is not meant for public usage.
+ *
+ * @private
+ * @experimental
+ */
+class XMLMultiPartUploadHelper implements MultiPartUploadHelper {
```

Comment:
should this go in a separate file?

Reply:
I can move this to a separate file the only problem is I will have to export it which means a user can potentially come along and import it and use it directly. I was attempting to avoid that situation. Depends what people think and whether or not we are ok with saying "don't use this directly".